src/pauli.jl

```diff
@@ -0,0 +1,229 @@
+    shape::Vector{Int}
+    bases::B
+    function PauliBasis(num_qubits::Int64)
+        return new{Tuple{(SpinBasis{1//2} for _ in 1:num_qubits)...}}([2 for _ in 1:num_qubits], Tuple(SpinBasis(1//2) for _ in 1:num_qubits))
+    end
+end
+==(pb1::PauliBasis, pb2::PauliBasis) = length(pb1.bases) == length(pb2.bases)
+
+"""
+Base class for Pauli transfer matrix classes.
+"""
+abstract type PauliTransferMatrix{B1<:Tuple{PauliBasis, PauliBasis}, B2<:Tuple{PauliBasis, PauliBasis}} end
+
+
+"""
+    DensePauliTransferMatrix(b, b, data)
+
+DensePauliTransferMatrix stored as a dense matrix.
+"""
+mutable struct DensePauliTransferMatrix{B1<:Tuple{PauliBasis, PauliBasis},
+                                        B2<:Tuple{PauliBasis, PauliBasis},
+                                        T<:Matrix{Float64}} <: PauliTransferMatrix{B1, B2}
+    basis_l::B1
+    basis_r::B2
+    data::T
+    function DensePauliTransferMatrix(basis_l::BL, basis_r::BR, data::T) where {BL<:Tuple{PauliBasis, PauliBasis},
+                                                                                BR<:Tuple{PauliBasis, PauliBasis},
+                                                                                T<:Matrix{Float64}}
+        if length(basis_l[1])*length(basis_l[2]) != size(data, 1) ||
+           length(basis_r[1])*length(basis_r[2]) != size(data, 2)
+            throw(DimensionMismatch())
+        end
+        new{BL, BR, T}(basis_l, basis_r, data)
+    end
+end
+
+PauliTransferMatrix(ptm::DensePauliTransferMatrix{B, B, Array{Float64, 2}}) where B <: Tuple{PauliBasis, PauliBasis} = ptm
+
+"""
+Base class for χ (process) matrix classes.
+"""
+abstract type ChiMatrix{B1<:Tuple{PauliBasis, PauliBasis}, B2<:Tuple{PauliBasis, PauliBasis}} end
+
+"""
+    DenseChiMatrix(b, b, data)
+
+DenseChiMatrix stored as a dense matrix.
+"""
+mutable struct DenseChiMatrix{B1<:Tuple{PauliBasis, PauliBasis},
+                              B2<:Tuple{PauliBasis, PauliBasis},
+                              T<:Matrix{Complex{Float64}}} <: PauliTransferMatrix{B1, B2}
+    basis_l::B1
+    basis_r::B2
+    data::T
+    function DenseChiMatrix(basis_l::BL, basis_r::BR, data::T) where {BL<:Tuple{PauliBasis, PauliBasis},
+                                                                                BR<:Tuple{PauliBasis, PauliBasis},
+                                                                                T<:Matrix{Complex{Float64}}}
+        if length(basis_l[1])*length(basis_l[2]) != size(data, 1) ||
+           length(basis_r[1])*length(basis_r[2]) != size(data, 2)
+            throw(DimensionMismatch())
+        end
+        new{BL, BR, T}(basis_l, basis_r, data)
+    end
+end
+
+DenseChiMatrix(chi_matrix::DenseChiMatrix{B, B, Array{Complex{Float64}, 2}}) where B <: Tuple{PauliBasis, PauliBasis} = chi_matrix
+
+"""
+    pauli_operators(num_qubits::Int64)
+
+Generate a matrix of basis vectors in the Pauli representation given a number
```

Comment:
Just learning about Julia's type system - is this actually a matrix?

It seems like matrix is Array{T,2} where T, while this seems to produce something like: Array{Array{Int64,1},1}.
Is an array of arrays formally equivalent to a matrix in julia?

Reply:
no, as you say, they are different. i am using matrix colloquially.